src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs

```diff
@@ -814,11 +814,14 @@ pub fn add_new_sub_level(
     if insert_sub_level_id == u64::MAX {
         return;
     }
-    if let Some(newest_level) = l0.sub_levels.last() {
+    let insert_pos = l0
```

Comment:
Any reason why there can be newly added sub level that has an older sub level id?

Reply:
> Any reason why there can be newly added sub level that has an older sub level id?

`init_with_parent_group` call in `hummock_version_ext.rs`

Reply:
In code some time before, we only use epoch as the sub level id. Just curious, in current code, besides epoch, what can be used as sub level id and how the sub level ids are generated?

Reply:
> In code some time before, we only use epoch as the sub level id. Just curious, in current code, besides epoch, what can be used as sub level id and how the sub level ids are generated?

Nothing. We only use epoch as the sub level id now.

Reply:
I just checked the code of `add_ssts_to_sub_level`. It only calls `add_new_sub_level` when the `sub_level_idx` is greater than the current number of sub levels, which means the semantic of the operation should be appending a new sub level to the end of the sub level lists. However, the modification in this PR will break this semantic, which may break the current semanic.
```
pub fn add_ssts_to_sub_level(
    l0: &mut OverlappingLevel,
    sub_level_idx: usize,
    insert_sub_level_id: u64,
    level_type: LevelType,
    insert_table_infos: Vec<SstableInfo>,
)
```

And I don't get the reason why in `init_with_parent_group`, we can have a new sub level with epoch older than the current latest epoch. Is there any example?

Reply:
> I just checked the code of `add_ssts_to_sub_level`. It only calls `add_new_sub_level` when the `sub_level_idx` is greater than the current number of sub levels, which means the semantic of the operation should be appending a new sub level to the end of the sub level lists. However, the modification in this PR will break this semantic, which may break the current semanic.
> 
> ```
> pub fn add_ssts_to_sub_level(
>     l0: &mut OverlappingLevel,
>     sub_level_idx: usize,
>     insert_sub_level_id: u64,
>     level_type: LevelType,
>     insert_table_infos: Vec<SstableInfo>,
> )
> ```
> 
> And I don't get the reason why in `init_with_parent_group`, we can have a new sub level with epoch older than the current latest epoch. Is there any example?

You can check the code below in `init_with_parent_group`, which is the answer to both your doubts.
```
                let target_l0 = cur_levels.l0.as_mut().unwrap();
                let mut target_level_idx = target_l0.sub_levels.len();
                for (idx, other) in target_l0.sub_levels.iter_mut().enumerate() {
                    if other.sub_level_id == sub_level.sub_level_id {
                        target_level_idx = idx;
                    }
                }
```

Reply:
I seem to get the idea. 

If we have the such need to insert a new sub level to a specific position, we'd better change to have two functions, named like `add_ssts_to_sub_level ` and `insert_new_sub_level` and meaning adding sst to an existing sub level in provided idx, and adding a new sub level to a specific position, respectively. In the `add_ssts_to_sub_level ` we won't call `insert_new_sub_level` any more, and in the `insert_new_sub_level` we can add a sanity check to ensure that the position(index) passed in is a valid position (after insertion, the sub level is sorted by sub level id).

Reply:
> I seem to get the idea.
> 
> If we have the such need to insert a new sub level to a specific position, we'd better change to have two functions, named like `add_ssts_to_sub_level ` and `insert_new_sub_level` and meaning adding sst to an existing sub level in provided idx, and adding a new sub level to a specific position, respectively. In the `add_ssts_to_sub_level ` we won't call `insert_new_sub_level` any more, and in the `insert_new_sub_level` we can add a sanity check to ensure that the position(index) passed in is a valid position (after insertion, the sub level is sorted by sub level id).

ok.